compile.c

```diff
@@ -2607,15 +2607,14 @@ iseq_set_sequence(rb_iseq_t *iseq, LINK_ANCHOR *const anchor)
     iseq_bits_t * mark_offset_bits;
     int code_size = code_index;
 
-    iseq_bits_t tmp[1] = {0};
     bool needs_bitmap = false;
 
-    if (ISEQ_MBITS_BUFLEN(code_index) == 1) {
-        mark_offset_bits = tmp;
-    }
-    else {
-        mark_offset_bits = ZALLOC_N(iseq_bits_t, ISEQ_MBITS_BUFLEN(code_index));
-    }
+    mark_offset_bits = ZALLOC_N(iseq_bits_t, ISEQ_MBITS_BUFLEN(code_index));
+
+    ISEQ_COMPILE_DATA(iseq)->iseq_encoded = (void *)generated_iseq;
+    ISEQ_COMPILE_DATA(iseq)->iseq_size = code_index;
+    ISEQ_COMPILE_DATA(iseq)->mark_bits.list = mark_offset_bits;
+    ISEQ_COMPILE_DATA(iseq)->is_single_mark_bit = false;
```

Comment:
Doesn't this render the single mark bit optimization useless? It's no longer conditionally allocating the `mark_offset_bits` but rather always allocating it and then freeing when it only needs a single word?

Reply:
The reason we're doing this is because `tmp` is stack allocated, and the garbage collector needs to access that memory if a GC happens to occur during complication. We decided to always malloc so that if a GC happens, it can access those mark bits. Later we free it. So the optimization isn't lost because overall process memory should remain the same. In other words, in some cases we will need to do an extra malloc. Unless we can point at stack allocated memory, which I don't think we can.

Reply:
We can take pointers to stack allocated memory, but we shouldn't need to do that as we can just point it to the `ISEQ_COMPILE_DATA(iseq)->mark_bits.single`.

Reply:
Looked at this with Aaron. We can't point `ISEQ_COMPILE_DATA(iseq)->mark_bits.single` to `tmp[0]` because it's a copy of the value (integer). Later in the function when the bits are marked, it'd be marking the value in `tmp[0]`, so the integer copy in `mark_bits.single` wouldn't get updated and the GC wouldn't see changes. So it's either this implementation or take a pointer to the stack memory. Are we misunderstanding what you're saying?

Reply:
I pushed a commit that should do this.